File: tools/elemfile.py
#!/usr/python3

#parse element input file

import math
import sys

class Elem:
	def __init__(self):
		self.m_fDamping = 0
		self.m_fArea = 0
		self.m_bBreak = False
		self.m_bGeom = True
		self.m_bSink = False
	
		# link syntax:
		#   -1: no link
		#   0: link to "infinity"
		#   >0: link to actual element
		self.m_iLink = -1
	
		# speaker will be inserted on the velocity link to the previous pressure element
		self.m_strSpeaker = ""
	
		# microphone definition. measures pressure of this pressure element
		self.m_strMic = ""
		
	
	def __str__(self):
		strDesc = "{element; area:" + str(self.m_fArea) + "; damping:" + str(self.m_fDamping)
		
		if self.m_strMic != "":
			strDesc += "; microphone \"" + self.m_strMic + "\""
		
		if self.m_strSpeaker != "":
			strDesc += "; speaker \"" + self.m_strSpeaker + "\""
		
		return strDesc + "}"

class Microphone:
	m_iElemID = 0
	def __str__(self):
		return "{microphone; element:" + str(self.m_iElemID) + "}"

class Speaker:
	m_iElemID = 0
	def __str__(self):
		return "{speaker; element:" + str(self.m_iElemID) + "}"

# returns dictionaries of objects
# return (dElems, dMics, dSpeakers, dx)

def scanElemFile(strFilename):
	aElems = []
	dMics = dict()
	dSpeakers = dict()

	f = open(strFilename, 'r')
	aLines = f.readlines()
	f.close()

#	delta X of the elements (length)
	dx = float("nan")

	iElem = -1

	for currLine in aLines:
		linetype = currLine[0]

#		print ("type of line", iLine, "is", linetype)

		if linetype == "e":
			iElem += 1
			elID = int(currLine[2:])
			
			if iElem != elID:
					print("ERROR: expected element ID to be", iElem, "not", elID)
#			print("element ID is", elID)
			elem = Elem()
			aElems.append( elem )
		elif linetype == "d":
			fDamping = float(currLine[2:])
			aElems[-1].m_fDamping = fDamping
		elif linetype == "A":
			fArea = float(currLine[2:])
			aElems[-1].m_fArea = fArea
		elif linetype == "b":	
			aElems[-1].m_bBreak = True
		elif linetype == "l":
			iLinkTo = int(currLine[2:])
			aElems[-1].m_iLink = iLinkTo
		elif linetype == "s":
			speakerID = currLine[3:-2]
			
			aElems[-1].m_strSpeaker = speakerID
			spkr = Speaker()
			
			spkr.m_iElemID = iElem
			dSpeakers[speakerID] = spkr
		elif linetype == "m":
			micID = currLine[3:-2]
			
			mic = Microphone()
			mic.m_iElemID = iElem
			dMics[micID] = mic
			aElems[-1].m_strMic = micID
		elif linetype == "g":
			aElems[-1].m_bGeom = True
		elif linetype == "i":
			aElems[-1].m_bSink = True
			
		elif linetype == "x":
			dx = float(currLine[3:])
		elif linetype == "#":
			pass
		elif currLine == "\n":
			pass
		else:
			print("type of line not recognized!", currLine)
	return (aElems, dMics, dSpeakers, dx)

def writeElemFile(strFilename, aElements, fDx):
	strFileContents = "# set global element length\n"
	strFileContents += "x " + str(fDx) + "\n"
	
	strFileContents += "# define elements\n"
	
	for iElem in range(len(aElements)):
		strFileContents += "e " + str(iElem) + "\n"
		elem = aElements[iElem]
		
		strFileContents += "A " + str(elem.m_fArea) + "\n"
		
		if elem.m_fDamping != 0:
			strFileContents += "d " + str(elem.m_fDamping) + "\n"
		if elem.m_iLink != -1:
			strFileContents += "l " + str(elem.m_iLink) + "\n"
		if elem.m_bBreak:
			strFileContents += "b\n"
		if elem.m_bGeom:
			strFileContents += "g\n"
		if elem.m_bSink:
			strFileContents += "i\n"
		if elem.m_strSpeaker != "":
			strFileContents += "s \"" + elem.m_strSpeaker + "\"\n"
		if elem.m_strMic != "":
			strFileContents += "m \"" + elem.m_strMic + "\"\n"
		
	f = open(strFilename, 'w')
	f.write(strFileContents)
	f.close()
		

if __name__ == "__main__":
	infile = sys.argv[1]
	dElems, dMics, dSpeakers, dx = scanElemFile(infile)

	for elem in aElems:
		print("elem", elem, dElems[elem])
	
	for mic in dMics.keys():
		print("mic", mic, dMics[mic])

	for speaker in dSpeakers.keys():
		print("speaker", speaker, dSpeakers[speaker])

	print("dx", dx)

File: lightsimu/elemFile.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "common.h"
#include "elemFile.h"

void
initializeElem(SElement* elem)
{
	elem->m_fDamping = 0;
	elem->m_fArea = 0;
	elem->m_bBreak = FALSE;
	elem->m_bGeometry = FALSE;

	// link syntax:
	//   -1: no link
	//   0: link to "infinity"
	//   >0: link to actual element
	elem->m_iLink = -1;

	// speaker will be inserted on the velocity link to the previous pressure element
	elem->m_strSpeaker = NULL;

	// microphone definition. measures pressure of this pressure element
	elem->m_strMic = NULL;
}

float
scanElemFile(const char* strFilename, SElement** ppElems, uint* pnElems)
{
	FILE * fp;
	char * line = NULL;
	size_t len = 0;
	size_t read;

	fp = fopen(strFilename, "r");
	if (fp == NULL)
		exit(EXIT_FAILURE);

	uint nElems = 0;
	while ((read = getline(&line, &len, fp)) != -1) {
		if(strncmp(line, "e", 1) == 0)
			nElems++;
	}

	rewind(fp);
	// allocate elements
	SElement* aElems = (SElement*) calloc(nElems, sizeof(SElement) );

	uint iElem = 0;
	for(; iElem < nElems; iElem++)
		initializeElem(& aElems[iElem]);

	iElem = -1;
	float fDeltaX = -1.0f;

	char buf[100];

	while ((read = getline(&line, &len, fp)) != -1) {
		if(strlen(line) < 1)
			continue;

		switch(line[0])
		{
		case 'e':
		{
			uint i;
			sscanf(line, "e %d", &i);

			iElem++;
			if(i != iElem)
				printf("scanFile: ERROR element number %d not expected (%d)", i, iElem);
			break;
		}
		case 'd':
			sscanf(line, "d %f", &aElems[iElem].m_fDamping);
			break;
		case 'i':
			sscanf(line, "i %f", &aElems[iElem].m_fInfiniteDamping);
			break;
		case 'A':
			sscanf(line, "A %f", &aElems[iElem].m_fArea);
			break;
		case 'l':
			sscanf(line, "l %d", &aElems[iElem].m_iLink);
			break;
		case 'b':
			aElems[iElem].m_bBreak = TRUE;
			break;
		case 's':
			sscanf(line, "s %s", buf);
			buf[strlen(buf)-1] = 0;
			aElems[iElem].m_strSpeaker = strdup(buf + 1);
			break;
		case 'm':
			sscanf(line, "m %s", buf);
			buf[strlen(buf)-1] = 0;
			aElems[iElem].m_strMic = strdup(buf + 1);
			break;
		case 'g':
			aElems[iElem].m_bGeometry = TRUE;
			break;
		case 'x':
			sscanf(line, "x %f", &fDeltaX);
			break;
		case '#':
			break;
		default:
			line[strlen(line) - 1] = 0;
			printf("scanFile: ERROR line %s not recognized\n", line);
		}
	}
	fclose(fp);
	*ppElems = aElems;
	*pnElems = nElems;

	return fDeltaX;
}

void
getComponents(const SElement* aElems, uint nElems,
		uint** ppSpeakers, uint* pnSpeakers,
		uint** ppMics, uint* pnMics)
{
	*ppSpeakers = NULL;
	*ppMics = NULL;

	*pnMics = 0;
	*pnSpeakers = 0;

	uint iElem = 0;
	for(; iElem < nElems; iElem ++)
	{
		if(aElems[iElem].m_strSpeaker )
		{
			(*pnSpeakers)++;
			*ppSpeakers = (uint*) realloc(*ppSpeakers, *pnSpeakers);
			(*ppSpeakers)[*pnSpeakers - 1] = iElem;
		}
		if(aElems[iElem].m_strMic )
		{
			(*pnMics)++;
			*ppMics = (uint*) realloc(*ppMics, *pnMics);
			(*ppMics)[*pnMics - 1] = iElem;
		}
	}
}
